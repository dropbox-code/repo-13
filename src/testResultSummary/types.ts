import { z } from "zod"

/**
 * Metadata describing a file that contains test results
 * for a specific test type
 */
const TestType = z.enum(["go" /*, "solidity"*/])
export const TestResultsFileMetadataSchema = z.object({
  testType: TestType,
  filePath: z.string().nonempty(),
})
export type TestResultsFileMetadata = z.infer<
  typeof TestResultsFileMetadataSchema
>

/**
 * The subschema for the test results that we want to handle
 * when generating the test result summary
 */
const handledTestStatuses = z.enum(["pass", "fail"])
type HandledTestStatuses = z.infer<typeof handledTestStatuses>
export type HandledTestResults = z.infer<typeof handledTestResultsSchema>
export const handledTestResultsSchema = z.object({
  Test: z.string().min(1).optional(),
  Action: handledTestStatuses,
  Elapsed: z.number().nonnegative(),
})

/**
 * A partial representation of the possible a single test log outputted by
 * https://pkg.go.dev/cmd/test2json
 *
 * Ideally, we would use `transform` to map the parsed test result to the
 * fields we want to use in the test result summary, but that breaks
 * discriminated unions.
 *
 * @see https://github.com/colinhacks/zod/issues/2315
 */
export type TestResult = z.infer<typeof TestResultSchema>
const TestResultSchema = z.discriminatedUnion("Action", [
  handledTestResultsSchema,
  z.object({
    Test: z.string().min(1).optional(),
    Action: z.literal("skip"),
    Elapsed: z.number().nonnegative(),
  }),
  z.object({
    Test: z.string().min(1),
    Action: z.literal("run"),
  }),
  z.object({
    Output: z.string().min(1),
    Action: z.literal("output"),
  }),
  z.object({
    Action: z.literal("start"),
  }),
  z.object({
    Action: z.literal("pause"),
  }),
  z.object({
    Action: z.literal("cont"),
  }),
])

/**
 * A representation of a file containing all test logs of
 * the JSONL format outputted by https://pkg.go.dev/cmd/test2json
 */
export const TestResultsSchema = z.array(TestResultSchema)

export interface MappedTestResult {
  name: string
  status: HandledTestStatuses
  statusInt?: number
  sha?: string
  elapsed: number
  jobRunId?: number
  jobName?: string
  repo?: string
}
