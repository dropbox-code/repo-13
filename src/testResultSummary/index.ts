import * as core from "@actions/core"
import * as fs from "fs"
import { TestResultsFileMetadata, MappedTestResult } from "./types"
import { parseGoTestResults } from "./parsers/golang"
import { ZodError } from "zod"
import { fromZodError } from 'zod-validation-error'

/**
 * Get the test results data from the file and prepare it for loki
 */
export function getTestResultSummary(
  fileMetadata: TestResultsFileMetadata,
): MappedTestResult[] {
  try {
    const fileData = fs.readFileSync(fileMetadata.filePath, "utf8")

    // format the data based on the test type, go is the first but solidity will be added soon
    switch (fileMetadata.testType) {
      case "go":
        return parseGoTestResults(fileData)
      default:
        throw Error("Unknown test type: " + fileMetadata.testType)
    }
  } catch (error) {
    if (error instanceof ZodError) {
      const validationError = fromZodError(error as ZodError)
      core.error(validationError)
    }
    core.error("Could not read the file: " + fileMetadata.filePath)
    throw error
  }
}
