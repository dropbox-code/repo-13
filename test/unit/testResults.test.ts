import { getTestResultSummary } from "@src/testResultSummary"
import {
  TestResultsFileMetadata,
  TestResultsFileMetadataSchema,
  MappedTestResult,
} from "@src/testResultSummary/types"

describe("Test Results Parsing", () => {
  it("should parse a go test results file input", async () => {
    const asInputString = '{"testType":"go","filePath":"./test/fixtures/testResults/go_test_results_input.json"}'
    const inputAsObject = JSON.parse(asInputString)
    const metadata: TestResultsFileMetadata = TestResultsFileMetadataSchema.parse(inputAsObject)
    const data: MappedTestResult[] = getTestResultSummary(metadata)
    expect(data).toMatchSnapshot()
  })
})
