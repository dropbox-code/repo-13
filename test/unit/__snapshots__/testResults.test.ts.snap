// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Results Parsing should parse a go test results file input 1`] = `
[
  {
    "elapsed": 506.4,
    "name": "TestMinResources5NodesEnvWithBlockscout",
    "status": "pass",
  },
  {
    "elapsed": 522.96,
    "name": "Test5NodesPlus2MiningGethsReorgEnv",
    "status": "fail",
  },
]
`;
